flowcontrol: remove test dependencies from go binaries

aojea opened the pull request:
remove testing dependencies on flowcontrol

NewFakeBackoff constructor is used to inject a clock for testing,
however, the function signature cast the input variable to the testing
clock type instead of the clock interface, injecting testing
depednencies into binaries.

Replace the testing type from the function parameter and use the clock.Clock
interface to keep backwards compatibility.

/kind bug
/kind cleanup
```release-note
NONE
```

/area code-organization

aojea commented:
/assign @liggitt @dims

fedebongio commented:
/triage accepted

aojea commented:
/assign @sttts @liggitt @dims

liggitt commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:
```diff
@@ -23,6 +23,12 @@ import (
 	testingclock "k8s.io/utils/clock/testing"
 )
 
+var NewFakeBackOff = NewBackOffWithClock
```
why export this? it's unusable by things outside this package

liggitt replied:
```sh
$ find . -type f | xargs bat
───────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: ./go.mod
───────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ module example.com
   2   │ 
   3   │ go 1.19
───────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
───────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: ./a/a_test.go
───────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ package a
   2   │ 
   3   │ import (
   4   │     "fmt"
   5   │     "testing"
   6   │ 
   7   │     "example.com/b"
   8   │ )
   9   │ 
  10   │ func TestA(t *testing.T) {
  11   │     fmt.Println(b.Exported)
  12   │ }
───────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
───────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: ./b/b_test.go
───────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ package b
   2   │ 
   3   │ var Exported = "Hello, world"
───────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
───────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: ./b/b.go
───────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ package b
───────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

$ go test ./a
# example.com/a [example.com/a.test]
a/a_test.go:11:16: undefined: b.Exported
FAIL	example.com/a [build failed]
FAIL
```

aojea replied:
I was trying to reduce the diff, nothing else, let me change it

aojea replied:
I removed it now

liggitt commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -49,23 +48,19 @@ type Backoff struct {
 	maxJitterFactor float64
 }
 
-func NewFakeBackOff(initial, max time.Duration, tc *testingclock.FakeClock) *Backoff {
```
you can't move these to a _test file and preserve the exported API for things outside this package

aojea replied:
is not preserved, is it?

liggitt replied:
"exported" symboles in _test files don't get exported outside the package

aojea replied:
yeah, but I mean, is a cosmetic problem, isn;t it?

aojea commented:
/hold
/test

liggitt commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -49,23 +48,19 @@ type Backoff struct {
 	maxJitterFactor float64
 }
 
-func NewFakeBackOff(initial, max time.Duration, tc *testingclock.FakeClock) *Backoff {
```
Even though it bothers me to keep the current name while taking an interface, doing so would preserve compatibility with existing callers. If we're going to try to improve go signature compatibility, I would probably switch this to `NewFakeBackOff(initial, max time.Duration, tc clock.Clock) *Backoff` rather than renaming the method

That preserves compatibility with all existing callers, at least

aojea replied:
yeah,

aojea committed:
flowcontrol: remove testing dependencies

The methods NewFakeClock were using a testing dependency as a parameter,
to avoid breaking compatibility and to remove this dependency, just use
the clock.Clock interface.

If we have to do it again most probable we have chosen other pattern and
for sure other names, but now is too late.

aojea commented:
/hold cancel

aojea commented:
/test pull-kubernetes-apidiff-client-go

aojea commented:
/test pull-kubernetes-apidiff-client-go

pacoxu commented:
> /test pull-kubernetes-apidiff-client-go

/hold
Should this be fixed or not(skip then)?

aojea commented:
> > /test pull-kubernetes-apidiff-client-go
> 
> /hold Should this be fixed or not(skip then)?

perfect, the job works as intended, let's take a look to the problems reported

```
Incompatible changes:
- ./util/flowcontrol.NewFakeBackOff: changed from func(time.Duration, time.Duration, *k8s.io/utils/clock/testing.FakeClock) *Backoff to func(time.Duration, time.Duration, k8s.io/utils/clock.Clock) *Backoff
- ./util/flowcontrol.NewFakeBackOffWithJitter: changed from func(time.Duration, time.Duration, *k8s.io/utils/clock/testing.FakeClock, float64) *Backoff to func(time.Duration, time.Duration, k8s.io/utils/clock.Clock, float64) *Backoff
```

we are changing from `func(time.Duration, time.Duration, *k8s.io/utils/clock/testing.FakeClock, float64)` to ` func(time.Duration, time.Duration, k8s.io/utils/clock.Clock, float64) *Backoff`

basically replacing `testing.FakeClock` with `clock.Clock`, that is an interface, despite is breaking, the consumers will not need to do anything (we have places in the codebase that use `NewFakeBackOff` and as we can see in the PR we didn't need to change them

```
grep -r NewFakeBackOff
pkg/controller/daemon/daemon_controller_test.go:                flowcontrol.NewFakeBackOff(50*time.Millisecond, 500*time.Millisecond, testingclock.NewFakeClock(time.Now())),
pkg/controller/daemon/daemon_controller_test.go:                flowcontrol.NewFakeBackOff(50*time.Millisecond, 500*time.Millisecond, testingclock.NewFakeClock(time.Now())),
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:  b := NewFakeBackOff(step, maxDuration, tc)
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:  b := NewFakeBackOff(step, maxDuration, tc)
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:  b := NewFakeBackOff(step, maxDuration, tc)
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:  b := NewFakeBackOff(step, maxDuration, tc)
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:          b := NewFakeBackOff(base, maxDuration, clock)
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:  b := NewFakeBackOff(step, maxDuration, tc)
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:  b := NewFakeBackOffWithJitter(initial, maxDuration, tc, maxJitterFactor)
staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:          b := NewFakeBackOff(base, maxDuration, clock)
staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:func NewFakeBackOff(initial, max time.Duration, tc *testingclock.FakeClock) *Backoff {
staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:func NewFakeBackOffWithJitter(initial, max time.Duration, tc *testingclock.FakeClock, maxJitterFactor float64) *Backoff {
staging/src/k8s.io/client-go/rest/request_test.go: 
```

/hold cancel

The job warned about a breaking change but we verified that consumers of the function will not be impacted, so this is good to go, thanks Paco for catching this

pacoxu commented:
/skip

aojea commented:
> /hold Should this be fixed or not(skip then)?

please keep doing this ALWAYS, the goal of the job is to make this changes after discussions like this and evaluating the pros and cons
